### PR TITLE
[1.20.6] Capability Factories PR 

### PR DESCRIPTION
**Pull request by RealMangorage**
This PR will allow Developers to register their Capability Providers through a factory, which the ItemStacks wiill then call to get the provider, allowing us to prevent the need to call AttachCapabilityEvent for each Instance creation of ItemStack.

I do plan on bringing this to other things as well, like Blocks and Entities.

### CODE REVIEW

**Comment by Choonster** [comment]
Is there a plan to allow item capabilities to store their data directly, like what was previously done with the `ICapabilityProvider` implementing `INBTSerializable`, or is it expected that they will manually store their data using something like a Data Component?

**Comment by LexManos** [comment]
It is expected that they manage it themselves with components.

File: src/main/java/net/minecraftforge/common/capabilities/factory/CapabilityFactoryManager.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    }
+
+    private record Key(Capability<?> cap, Direction side) {}
+    private record Exposed(Capability<?> cap, Direction[] dirs) {}
+    private record Instance<I>(Function<I, LazyOptional<?>> func, List<Exposed> exposed) {}
+
+    private final static class CapabilityFactoryProvider<I> implements ICapabilityProvider {
+        private final CapabilityFactoryManager<I, ?> manager;
+        private final Map<CapabilityFactoryManager.Key, LazyOptional<?>> caps = new HashMap<>();
+        private final I instance;
```
Comment (PaintNinja) [inline src/main/java/net/minecraftforge/common/capabilities/factory/CapabilityFactoryManager.java:58]: Could this be a record?

File: src/main/java/net/minecraftforge/event/RegisterCapabilityFactoriesEvent.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.eventbus.api.GenericEvent;
+import org.apache.logging.log4j.util.TriConsumer;
+
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.function.Function;
+
+public class RegisterCapabilityFactoriesEvent<I, T> extends Event {
```
Comment (PaintNinja) [inline src/main/java/net/minecraftforge/event/RegisterCapabilityFactoriesEvent.java:14]: Missing javadocs

File: src/main/java/net/minecraftforge/common/capabilities/factory/CapabilityFactoryManager.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import net.minecraftforge.event.RegisterCapabilityFactoriesEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class CapabilityFactoryManager<I, T> implements ICapabilityFactoryManager<I> {
```
Comment (PaintNinja) [inline src/main/java/net/minecraftforge/common/capabilities/factory/CapabilityFactoryManager.java:17]: Missing javadocs. If internal, missing internal annotation

**Comment by Jonathing** [comment]
It was discussed in the Discord earlier, but it's likely that this will be put off until the beta period of Forge 1.21. I can also help with the porting process for this PR if you need it.

**Comment by PaintNinja** [comment]
Marking as a draft as this PR is incomplete. @RealMangorage once you've finished working on this, please mark it as ready for review to continue.

**Comment by Jonathing** [comment]
Superseded by #10118.

**Pull request closed without merging**